docs/Dialects/FIRRTL/RationaleFIRRTL.md

```diff
@@ -142,6 +142,154 @@ by Chisel typically look like `_T_12`, and names generated by the SFC look like
+
+``` verilog
+module Foo(
+  input  a,
+  output b);
+
+  wire named = a;
+  assign b = named;
+endmodule
+```
+
+When compiled with `-preserve-values=all` this produces the following Verilog
+that preserves all nodes, regardless of name:
+
+``` verilog
+module Foo(
+  input  a,
+  output b);
+
+  wire named = a;
+  wire _unnamed = named;
+  assign b = _unnamed;
+endmodule
+```
+
+Design teams are expected to use `-O=debug` debuggabilty.  Verification teams
+and downstream tools are expected to use/consume `-O=release`.
+
+This split of two different Verilog outputs initially created reproducibility
+problems that CIRCT has attempted to solve with a guarantee of stable
+randomization.  Consider a situation where a verification team trips an
+assertion failure using a release build with a particular seed.  Because release
+Verilog is highly optimized and difficult to debug, they want to switch to a
+debug build.  If the release build seed does not reproduce the failure in debug
+mode, the verification team needs to search for a failure.  This proved to be a
+drag on Chisel users.  Towards alleviating this, CIRCT will now guarantee that
+registers in debug or release mode will be randomized to the same value for the
+same seed.
+
+It is important to note that the debug/release split was born out of our
+inability to reconcile the goals at the top of this section.  Discussion in the
+subsequent section involves approaches to unify these two approaches.
+
+#### Alternative Approaches to Name Preservation Modes
+
+The following alternatives were considered instead of the debug/release
+solution.
+
+First, we originally considered having CIRCT create "debug modules" that
+included all named signals in the design.  An instance of this debug module
+would then be instantiated, via a SystemVerilog `bind` statement, inside the
+original module.  This was an early suggestion.  However, a concern of users of
+any "debug module" is that the debug module would not show usages of the named
+signals.  E.g., the example circuit shown above would compile to something like:
+
+
+``` verilog
+module Foo_debug(
+  input _0;
+
+  wire named = _0;
+endmodule
+
+bind Foo Foo_debug Foo_debug (
+  ._0(a)
+);
+
+module Foo(
+  input  a,
+  output b);
+
+  assign b = a;
+endmodule
+```
+
+The main concern is that while users can see the value of `named` in a waveform,
+they cannot trace back its usage in the computation of port `b` in module `Foo`.
+
+This approach may be revisited in the future as it provides benefits of unifying
+debug and release builds into a single release build with run-time debugging
+information that can be bound in.
+
+Second, a single build that always preserved names was considered.  At the time,
+this introduced long Verilog compilation and simulation times.  We were not able
+to discern an optimization design point which balanced the needs of
+debuggability with compilation and simulation performance.  This does not mean
```

Comment:
There was another concern here, which was, IIRC, the real showstopper at the time. In name preservation mode (-preserve-values=named), the preservation of certain named values disabled certain optimizations that were required for "correctness". I believe the specific issue was ports were not removed, because there were named signals that were assigned from the ports, even though they were unused. Because of this, I believe verification had coverage holes. We don't want optimizations to be required for correctness, but we were also trying to replace an existing FIRRTL compiler at the time, and had inherited this expectation.

We looked at another angle, where in name preservation mode we could remove named signals in certain scenarios like the above, but we didn't find a satisfying implementation. Named wires are preserved by giving them a symbol, just like a don't touch, so there's no safe way to remove named wire in the SV dialect. We tried other ways to indicate the wire should exist just for its name, but this was special cased and hacky.

Since then, other work in this realm has happened. For example, grand central changed to not use don't touch, and the IMDCE pass was developed to address port removal. Similar work might be able to revive the name preservation approach, but with certain optimizations enabled. Or that might remain a dead end.

I'm not sure if it's worth mentioning the above, but I thought I'd mention it.

Reply:
Great background. I'll roll this into the current doc. 👍

Reply:
Thanks for writing this up, Mike. This jogged my memory on some of the issues and I synthesized this into some of the existing sections.